clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -3058,23 +3225,40 @@ scheduler_tmpl_t<RecordType, ReaderType>::pick_next_input(output_ordinal_t outpu
                 }
                 if (index != INVALID_INPUT_ORDINAL) {
                     // We found a direct switch target above.
-                } else if (ready_queue_empty() && blocked_time == 0) {
-                    if (prev_index == INVALID_INPUT_ORDINAL)
-                        return eof_or_idle(output, need_lock, prev_index);
-                    auto lock =
-                        std::unique_lock<mutex_dbg_owned>(*inputs_[prev_index].lock);
-                    // If we can't go back to the current input, we're EOF or idle.
-                    // TODO i#6959: We should go the EOF/idle route if
-                    // inputs_[prev_index].unscheduled as otherwise we're ignoring its
-                    // unscheduled transition: although if there are no other threads at
-                    // all (not just an empty queue) this turns into the eof_or_idle()
-                    // all-unscheduled scenario.  Once we have some kind of early exit
-                    // option we'll add the unscheduled check here.
-                    if (inputs_[prev_index].at_eof) {
-                        lock.unlock();
-                        return eof_or_idle(output, need_lock, prev_index);
-                    } else
-                        index = prev_index; // Go back to prior.
+                }
+                // XXX: We're grabbing the output runqueue lock 3x here:
+                // ready_queue_empty(), set_cur_input()'s add_to_ready_queue(),
+                // and pop_from_ready_queue().  Should we change the lock
+                // scheme to be in the caller and grab once here?
```

Comment:
You could have these methods take an optional pointer to the mutex as an indication the caller already grabbed the lock, so they'd only acquire a lock themselves if the pointer is null.  It would be using the pointer argument like a bool, but with type checking.

Reply:
Having the lock held across the code in between may be complex though: added that to the comment.